spec/src/main/asciidoc/app-programming-model.adoc

```diff
@@ -579,8 +580,12 @@ Gauge<Double> hitPercentage = new Gauge<Double>() {
 };
 ----
 
-When a metric injected field is annotated, the implementation must provide the registered metric with the given name (using the <<annotated-naming-convention>>) if the metric already exist. If no metric exists with the given name then the implementation must produce and register the requested metric.
-`@Metric` can only be used on injected fields of type `Meter`, `Timer`, `Counter`, and `Histogram`.
```

Comment:
I think there's confusion here.  the statement about only being able to inject fields of certain types is still true.  it applies to the example below.  You can't inject a Gauge via an `@Metric` since a gauge is a method callback.

Reply:
I'm not sure I understand your point. You can inject a Gauge, it just behaves a bit differently, as explained in the following paragraph. The main point of this PR is to fix the issue where the spec says you can't inject a Gauge while you actually can (the TCK indeed verifies that).

Reply:
I think the original text might've wanted to clarify that the @Inject and @Metric wouldn't create/produce and _register_ a gauge for you after just saying the implementation _would_ "produce and register the requested metrics" if it didn't exist.

The new text sort of does imply this isn't the case. But maybe it might clear it up if we explicitly say Gauges aren't produced and registered, but rather have a proxy gauge etc etc.

Reply:
Injection point resolution happens at application boot time, as well as registering metrics from annotations (which is optional for gauges). It is not defined which of these two operations must happen first. So depending purely on implementation details, when the `@Inject` field is being resolved, the gauge might or might not exist. It can also happen that the gauge will be registered programmatically later at runtime, we have no way of knowing that when we're performing the boot-time injection point resolution.

The new text says that when there is an `@Inject @Metric` then the relevant metric should be registered in the registry, with the exception of gauges, because we don't know in advance whether there will be a gauge registered under that ID, and we can't just register "any" gauge until we have an implementation. So it can happen that there is a `@Inject @Metric Gauge gauge` field, but no actual gauge will ever be registered in the metric registry. 

I feel the new text kinda reflects this. But perhaps I just don't understand what your point is, feel free to suggest a particular different wording, that might be better than trying to explain what is wrong with the current wording.

Reply:
It was just more of a comment on how the original text sent the wrong message by not being explicit enough. Hence the confusion that was caused.

The new text does do a better job explaining what's going on. But maybe a more explicit declaration? (i.e see below).
However, otherwise I'm okay with the changes :D.

```
Gauges are an exception to this rule, because it could happen that an annotated gauge is not registered yet when the reference to it is being injected
```
to
```
Gauges are an exception to this rule.  The implementation will not provide a new Gauge metric, but rather attempt to inject one even if the defined Gauge metric has not been registered yet.
```

Reply:
This was just misunderstanding on my part.  I'm happy with the originally proposed changes related to gauge.  Sorry for making work -- it looks good.